update docs

sangyx opened the pull request:
Hi, I have updated the docs. While I found my setting `doctest_global_setup` in docs/conf.py does not work. Could you help check? Many thanks!

Besides, I want to change the diff function form of JVP from `lambda g, ans, x, y: ` to `lambda ans, x, y: lambda g: `, so as to separate the calculation of the result from the derivation process and solve the limitations of grad_variables, jacobian, and higher derivative. But this may not be done before September, so I will submit the GSoC result document first.

hameerabbasi commented:
You need to add `udiff` and so on to the `doctest_namespace`. https://docs.pytest.org/en/latest/doctest.html#doctest-namespace-fixture

hameerabbasi commented:
Here's an example: https://github.com/Quansight-Labs/uarray/blob/master/uarray/conftest.py

sangyx commented:
Thanks! I have added namespaces:
```python
doctest_namespace["ua"] = uarray
doctest_namespace["np"] = unumpy
doctest_namespace["onp"] = numpy
doctest_namespace["udiff"] = udiff
doctest_namespace["numpy_backend"] = numpy_backend
doctest_namespace["broadcast"] = udiff._jvp_diffs.broadcast
```

But the test throws the error `NameError: name 'broadcast' is not defined`. I could not find a solution in network.
```python
        >>> with ua.set_backend(udiff.DiffArrayBackend(numpy_backend, mode="jvp"), coerce=True):
        ...
        ...    x1 = np.array([2])
        ...    x2 = np.array([5])
        ...    y = np.log(x1) + x1 * x2 - np.sin(x2)
        ...    x1_diff = y.to(x1)
        ...    print(onp.allclose(x1_diff, [5.5]))
        True
```

hameerabbasi commented on conftest.py:
```diff
@@ -0,0 +1,27 @@
+    doctest_namespace["onp"] = numpy
+    doctest_namespace["udiff"] = udiff
+    doctest_namespace["numpy_backend"] = numpy_backend
+    doctest_namespace["broadcast"] = udiff._jvp_diffs.broadcast
```
Do we need this name anywhere?

hameerabbasi replied:
It seems we don't use it in the doctests, so we can remove this line.

hameerabbasi commented:
This is strange, does running the example work otherwise?

sangyx commented:
> This is strange, does running the example work otherwise?

If we use functions that don't not involve `broadcast`, it will work fine.

hameerabbasi commented:
Does `broadcast` work outside doctests?

sangyx commented:
> Does `broadcast` work outside doctests?

No. It is strange that I also used the `broadcast` in `defjvp`:
```python
    >>> defjvp(
    ...     np.subtract,
    ...     lambda g, ans, x, y: broadcast(g, ans),
    ...     lambda g, ans, x, y: broadcast(-g, ans),
    ... )
```
But it can through the test.

hameerabbasi commented:
Can you try replacing it with `np.broadcast`?

sangyx commented:
I think the error is caused by the doctest in defjvp using broadcast, but the error message keeps pointing to `JVPDiffArray.to`. I have replaced it with another example.

Besides, I have replaced the `broadcast` with `np.broadcast_to`

hameerabbasi commented:
Thanks, @sangyx!
